client/app/components/data-sources/EditDataSourceForm.jsx

```diff
@@ -0,0 +1,70 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { DataSource } from '@/services/data-source';
+import DynamicForm from '@/components/dynamic-form/DynamicForm';
+import helper from '@/components/dynamic-form/dynamicFormHelper';
+
+const HELP_LINKS = {
+  athena: 'https://redash.io/help/data-sources/amazon-athena-setup',
+  bigquery: 'https://redash.io/help/data-sources/bigquery-setup',
+  url: 'https://redash.io/help/data-sources/querying-urls',
+  mongodb: 'https://redash.io/help/data-sources/mongodb-setup',
+  google_spreadsheets: 'https://redash.io/help/data-sources/querying-a-google-spreadsheet',
+  google_analytics: 'https://redash.io/help/data-sources/google-analytics-setup',
+  axibasetsd: 'https://redash.io/help/data-sources/axibase-time-series-database',
+  results: 'https://redash.io/help/user-guide/querying/query-results-data-source',
+};
+
+export default function EditDataSourceForm({ dataSource, type, onSuccess, ...props }) {
+  const selectedType = type.type;
+  const fields = helper.getFields(type.configuration_schema, dataSource);
+
+  const handleSubmit = (values, successCallback, errorCallback) => {
+    helper.updateTargetWithValues(dataSource, values);
+    dataSource.$save(
+      (data) => {
+        successCallback('Saved.');
+        onSuccess(data);
+      },
+      (error) => {
+        if (error.status === 400 && 'message' in error.data) {
```

Comment:
You have to check if `error.data` exists - this callback is called on any error (it's actually `dataSource.$save(...).$promise.catch()`), and `error` may be an `Error` object which does not have `data` property - in this case `in` will throw an exception.